packages/graph/hash_graph/lib/graph/src/store/query/mod.rs

```diff
@@ -147,106 +167,177 @@ impl Expression {
     }
 }
 
+#[derive(Debug)]
+pub struct ExpressionError;
+
+impl fmt::Display for ExpressionError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.write_str("evaluation of expression failed")
+    }
+}
+
+impl Error for ExpressionError {}
+
 impl Expression {
-    // TODO: Implement error handling
-    //   see https://app.asana.com/0/0/1202884883200968/f
-    #[expect(clippy::missing_panics_doc, reason = "Error handling not applied yet")]
-    pub fn evaluate<'a, R, C>(&'a self, resolver: &'a R, context: &'a C) -> BoxFuture<Literal>
+    pub fn evaluate<'a, R, C>(
+        &'a self,
+        resolver: &'a R,
+        context: &'a C,
+    ) -> BoxFuture<Result<Literal, ExpressionError>>
     where
         R: Resolve<C> + Sync,
         C: Sync,
     {
         async move {
-            match self {
+            Ok(match self {
                 Expression::Eq(expressions) => {
                     for expression in expressions.windows(2) {
                         if !compare(
-                            &expression[0].evaluate(resolver, context).await,
-                            &expression[1].evaluate(resolver, context).await,
-                        ) {
-                            return Literal::Bool(false);
+                            &expression[0].evaluate(resolver, context).await?,
+                            &expression[1].evaluate(resolver, context).await?,
```

Comment:
This can use the related errors capabilities of `error-stack` to provide better diagnostics (fail-slow)

Reply:
That's a good point!
I played around a little bit (and also added it to `Expression::All`) and this results in
```
Could not query from store
├╴lib/graph/src/store/postgres/ontology/property_type/mod.rs:112:22
│
╰┬▶ evaluation of expression failed
 │  ├╴lib/graph/src/store/query/mod.rs:79:17
 │  ├╴unsupported operation: 4 == true
 │  ╰╴backtrace with 106 frames (1)
 │
 ╰▶ evaluation of expression failed
    ├╴lib/graph/src/store/query/mod.rs:79:17
    ├╴unsupported operation: "string" == 10.0
    ╰╴backtrace with 111 frames (2)
```

The ergonomics of creating those reports need to be improved (I know about your tuple suggestion), but that's another story 🙂

Currently, we don't use features from `error-stack` after `0.1` (besides `into_report()` over `report()` because of the toolchain fix). We, however, will switch to `0.2` ASAP.